Flaky integration test `test_storage_rabbitmq`

https://s3.amazonaws.com/clickhouse-test-reports/66532/ab2380a9f09c0ce89ea31311ab01acd87e677ce8/integration_tests__tsan__[6_6].html

Al rabbitmq tests have failed or thrown an error. Checking https://s3.amazonaws.com/clickhouse-test-reports/66532/ab2380a9f09c0ce89ea31311ab01acd87e677ce8/integration_tests__tsan__[6_6]/integration_run_parallel4_0.tar.zst

I see:

```
1065:2024-10-24 17:02:20.281556+00:00 [error] <0.1566.0> Channel error on connection <0.1557.0> (172.16.4.4:48356 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 1:
1066:2024-10-24 17:02:20.281556+00:00 [error] <0.1566.0> operation exchange.declare caused a channel exception precondition_failed: inequivalent arg 'durable' for exchange 'f' in vhost '/': received 'true' but current is 'false'
1222:2024-10-24 17:04:06.936672+00:00 [error] <0.2256.0> Channel error on connection <0.2247.0> (172.16.4.1:63724 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 1:
1223:2024-10-24 17:04:06.936672+00:00 [error] <0.2256.0> operation queue.declare caused a channel exception not_found: no queue 'rabbit_queue_drop' in vhost '/'
1695:2024-10-24 17:06:00.398984+00:00 [error] <0.6314.0> Channel error on connection <0.5754.0> (172.16.4.4:50816 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 4:
1696:2024-10-24 17:06:00.398984+00:00 [error] <0.6314.0> operation queue.delete caused a channel exception precondition_failed: queue '1_many_consumers' in vhost '/' in use
1698:2024-10-24 17:06:00.632674+00:00 [error] <0.6328.0> Channel error on connection <0.5798.0> (172.16.4.4:40384 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 4:
1699:2024-10-24 17:06:00.632674+00:00 [error] <0.6328.0> operation queue.delete caused a channel exception precondition_failed: queue '1_many_consumers' in vhost '/' in use
1701:2024-10-24 17:06:00.858751+00:00 [error] <0.6342.0> Channel error on connection <0.5828.0> (172.16.4.4:40388 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 4:
1702:2024-10-24 17:06:00.858751+00:00 [error] <0.6342.0> operation queue.delete caused a channel exception precondition_failed: queue '1_many_consumers' in vhost '/' in use
1974:2024-10-24 17:07:21.832520+00:00 [error] <0.2646.0> closing AMQP connection <0.2646.0> (172.16.4.1:60108 -> 172.16.4.2:5672, duration: '3M, 0s'):
1975:2024-10-24 17:07:21.832520+00:00 [error] <0.2646.0> missed heartbeats from client, timeout: 60s
2306:2024-10-24 17:10:49.510519+00:00 [error] <0.9585.0> Channel error on connection <0.9148.0> (172.16.4.4:56446 -> 172.16.4.2:5672, vhost: '/', user: 'root'), channel 4:
2307:2024-10-24 17:10:49.510519+00:00 [error] <0.9585.0> operation queue.delete caused a channel exception precondition_failed: queue '1_random' in vhost '/' in use
2353:2024-10-24 17:11:07.793439+00:00 [error] <0.9729.0> Error on AMQP connection <0.9729.0> (172.16.4.1:58558 -> 172.16.4.2:5672, vhost: '/', user: 'root', state: running), channel 0:
2354:2024-10-24 17:11:07.793439+00:00 [error] <0.9729.0>  operation none caused a connection exception connection_forced: "broker forced connection closure with reason 'shutdown'"
2355:2024-10-24 17:11:07.793444+00:00 [error] <0.9701.0> Error on AMQP connection <0.9701.0> (172.16.4.4:53246 -> 172.16.4.2:5672, vhost: '/', user: 'root', state: running), channel 0:
2356:2024-10-24 17:11:07.793444+00:00 [error] <0.9701.0>  operation none caused a connection exception connection_forced: "broker forced connection closure with reason 'shutdown'"
2357:2024-10-24 17:11:07.793451+00:00 [error] <0.9663.0> Error on AMQP connection <0.9663.0> (172.16.4.4:53230 -> 172.16.4.2:5672, vhost: '/', user: 'root', state: running), channel 0:
2358:2024-10-24 17:11:07.793451+00:00 [error] <0.9663.0>  operation none caused a connection exception connection_forced: "broker forced connection closure with reason 'shutdown'"
2671:2024-10-24 17:11:16.140095+00:00 [debug] <0.296.0> ra_log_segment_writer: error sending ra_log_event to: RABBIT6KG0VW9M7ZBN. Error: No Pid
3246:2024-10-24 17:14:21.558924+00:00 [error] <0.737.0> closing AMQP connection <0.737.0> (172.16.4.4:60680 -> 172.16.4.2:5672, duration: '3M, 0s'):
3247:2024-10-24 17:14:21.558924+00:00 [error] <0.737.0> missed heartbeats from client, timeout: 60s
3291:2024-10-24 17:27:20.168729+00:00 [error] <0.803.0> closing AMQP connection <0.803.0> (172.16.4.4:44368 -> 172.16.4.2:5672, duration: '3M, 0s'):
3292:2024-10-24 17:27:20.168729+00:00 [error] <0.803.0> missed heartbeats from client, timeout: 60s
3303:2024-10-24 17:37:36.696945+00:00 [error] <0.1012.0> closing AMQP connection <0.1012.0> (172.16.4.4:37974 -> 172.16.4.2:5672, duration: '3M, 0s'):
3304:2024-10-24 17:37:36.696945+00:00 [error] <0.1012.0> missed heartbeats from client, timeout: 60s
```

I think it looks a setup/configuration issue for rabbitmq, but I can't pinpoint the root cause.

> Still flaky

From https://s3.amazonaws.com/clickhouse-test-reports/0/250a91aea4ca6d7202eb4a5c7a677a63a1a5423a/integration_tests__asan__[4_4]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel7_0.log

```
[gw1] [ 93%] FAILED test_storage_rabbitmq/test.py::test_rabbitmq_sharding_between_queues_publish Killed
```

😢

> It is killed because of the timeout (1h) likely

Yep, that completely checks

```
Mon, 03 Feb 2025 18:28:27 GMT Running pytest container as: 'docker run   --rm --name clickhouse_integration_tests_gym4z9
...
Mon, 03 Feb 2025 18:43:33 GMT test_storage_rabbitmq/test.py::test_row_based_formats 
Mon, 03 Feb 2025 19:28:26 GMT [gw1] [100%] FAILED test_storage_rabbitmq/test.py::test_row_based_formats Killed
```

By the time it times out, several tests have already failed, though. It's killed because the test execution is hanged somehow, but what we need to find is the root cause for those failures, which hopefully is also be the reason behind the frozen test 👀

I've seen several instances where `test_storage_rabbitmq/test.py::test_rabbitmq_no_connection_at_startup_2` fails due to timeout in the call to `rabbitmq_cluster.unpause_container("rabbitmq1")`:

- https://s3.amazonaws.com/clickhouse-test-reports/REFs/master/da267040c198a5bf40e29f97cf0c96674ba386a1//integration_tests_release_3_4/integration_run_parallel7_0.log
- https://s3.amazonaws.com/clickhouse-test-reports/REFs/master/638e7fef64ef05ea76aace6160b022531b10f02f//integration_tests_tsan_3_6/integration_run_parallel3_0.log

```
=================================== FAILURES ===================================
___________________ test_rabbitmq_no_connection_at_startup_2 ___________________
[gw0] linux -- Python 3.10.12 /usr/bin/python3

rabbitmq_cluster = <helpers.cluster.ClickHouseCluster object at 0x7f5810ae0220>

    def test_rabbitmq_no_connection_at_startup_2(rabbitmq_cluster):
        instance.query(
            """
            CREATE TABLE test.cs (key UInt64, value UInt64)
                ENGINE = RabbitMQ
                SETTINGS rabbitmq_host_port = 'rabbitmq1:5672',
                         rabbitmq_exchange_name = 'cs',
                         rabbitmq_format = 'JSONEachRow',
                         rabbitmq_num_consumers = '5',
                         rabbitmq_flush_interval_ms=1000,
                         rabbitmq_max_block_size=100,
                         rabbitmq_row_delimiter = '\\n';
            CREATE TABLE test.view (key UInt64, value UInt64)
                ENGINE = MergeTree
                ORDER BY key;
            CREATE MATERIALIZED VIEW test.consumer TO test.view AS
                SELECT * FROM test.cs;
        """
        )
        instance.query("DETACH TABLE test.cs")
        rabbitmq_cluster.pause_container("rabbitmq1")
        instance.query("ATTACH TABLE test.cs")
>       rabbitmq_cluster.unpause_container("rabbitmq1")

test_storage_rabbitmq/test.py:2463: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers/cluster.py:3207: in unpause_container
    subprocess_check_call(self.base_cmd + ["unpause", instance_name])
helpers/cluster.py:238: in subprocess_check_call
    return run_and_check(args, detach=detach, nothrow=nothrow, **kwargs)
helpers/cluster.py:131: in run_and_check
    res = subprocess.run(
/usr/lib/python3.10/subprocess.py:505: in run
    stdout, stderr = process.communicate(input, timeout=timeout)
/usr/lib/python3.10/subprocess.py:1154: in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
/usr/lib/python3.10/subprocess.py:2022: in _communicate
    self._check_timeout(endtime, orig_timeout, stdout, stderr)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Popen: returncode: -9 args: ['docker', 'compose', '--env-file', '/ClickHous...>
endtime = 36556.1753815, orig_timeout = 300, stdout_seq = [], stderr_seq = []
skip_check_and_raise = False

    def _check_timeout(self, endtime, orig_timeout, stdout_seq, stderr_seq,
                       skip_check_and_raise=False):
        """Convenience for checking if a timeout has expired."""
        if endtime is None:
            return
        if skip_check_and_raise or _time() > endtime:
>           raise TimeoutExpired(
                    self.args, orig_timeout,
                    output=b''.join(stdout_seq) if stdout_seq else None,
                    stderr=b''.join(stderr_seq) if stderr_seq else None)
E           subprocess.TimeoutExpired: Command '['docker', 'compose', '--env-file', '/ClickHouse/tests/integration/test_storage_rabbitmq/_instances-0-gw0/.env', '--project-name', 'rootteststoragerabbitmq-gw0', '--file', '/ClickHouse/tests/integration/test_storage_rabbitmq/_instances-0-gw0/instance/docker-compose.yml', '--file', '/ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_rabbitmq.yml', '--file', '/ClickHouse/tests/integration/test_storage_rabbitmq/_instances-0-gw0/instance2/docker-compose.yml', '--file', '/ClickHouse/tests/integration/test_storage_rabbitmq/_instances-0-gw0/instance3/docker-compose.yml', 'unpause', 'rabbitmq1']' timed out after 300 seconds

/usr/lib/python3.10/subprocess.py:1198: TimeoutExpired
```

--- a/docker/test/integration/runner/misc/rabbitmq/rabbitmq.conf
+++ b/docker/test/integration/runner/misc/rabbitmq/rabbitmq.conf
@@ -14,4 +14,4 @@ ssl_options.cacertfile = /etc/rabbitmq/ca-cert.pem
 ssl_options.certfile = /etc/rabbitmq/server-cert.pem
 ssl_options.keyfile = /etc/rabbitmq/server-key.pem
 
-vm_memory_high_watermark.absolute = 4GB
+vm_memory_high_watermark.absolute = 16GB

--- a/tests/integration/helpers/cluster.py
+++ b/tests/integration/helpers/cluster.py
@@ -276,7 +276,7 @@ def check_postgresql_java_client_is_available(postgresql_java_client_id):
     return p.returncode == 0
 
 
-def check_rabbitmq_is_available(rabbitmq_id, cookie, timeout=90):
+def run_rabbitmqctl(rabbitmq_id, cookie, command, timeout=90):
     try:
         subprocess.check_output(
             docker_exec(
@@ -284,24 +284,28 @@ def check_rabbitmq_is_available(rabbitmq_id, cookie, timeout=90):
                 f"RABBITMQ_ERLANG_COOKIE={cookie}",
                 rabbitmq_id,
                 "rabbitmqctl",
-                "await_startup",
+                command,
             ),
             stderr=subprocess.STDOUT,
             timeout=timeout,
         )
-        return True
     except subprocess.CalledProcessError as e:
         # Raised if the command returns a non-zero exit code
         error_message = (
-            f"RabbitMQ startup failed with return code {e.returncode}. "
+            f"rabbitmqctl {command} failed with return code {e.returncode}. "
             f"Output: {e.output.decode(errors='replace')}"
         )
         raise RuntimeError(error_message)
     except subprocess.TimeoutExpired as e:
         # Raised if the command times out
         raise RuntimeError(
-            f"RabbitMQ startup timed out. Output: {e.output.decode(errors='replace')}"
+            f"rabbitmqctl {command} timed out. Output: {e.output.decode(errors='replace')}"
         )
+
+
+def check_rabbitmq_is_available(rabbitmq_id, cookie, timeout=90):
+    run_rabbitmqctl(rabbitmq_id, cookie, "await_startup", timeout)
+    return True
 
 
 def rabbitmq_debuginfo(rabbitmq_id, cookie):
@@ -2386,6 +2390,18 @@ class ClickHouseCluster:
                 time.sleep(0.5)
 
         raise RuntimeError("Cannot wait RabbitMQ container")
+
+    def stop_rabbitmq_app(self, timeout=120):
+        run_rabbitmqctl(self.rabbitmq_docker_id, self.rabbitmq_cookie, "stop_app", timeout)
+
+    def start_rabbitmq_app(self, timeout=120):
+        run_rabbitmqctl(self.rabbitmq_docker_id, self.rabbitmq_cookie, "start_app", timeout)
+        self.wait_rabbitmq_to_start()
+
+    def reset_rabbitmq(self, timeout=120):
+        self.stop_rabbitmq_app()
+        run_rabbitmqctl(self.rabbitmq_docker_id, self.rabbitmq_cookie, "reset", timeout)
+        self.start_rabbitmq_app()
 
     def wait_nats_is_available(self, max_retries=5):
         retries = 0

--- a/tests/integration/test_storage_rabbitmq/test.py
+++ b/tests/integration/test_storage_rabbitmq/test.py
@@ -13,7 +13,7 @@ import pytest
 from google.protobuf.internal.encoder import _VarintBytes
 
 from helpers.client import QueryRuntimeException
-from helpers.cluster import ClickHouseCluster, check_rabbitmq_is_available
+from helpers.cluster import ClickHouseCluster
 from helpers.test_tools import TSV
 
 
@@ -71,9 +71,6 @@ def rabbitmq_cluster():
     try:
         cluster.start()
         logging.debug("rabbitmq_id is {}".format(instance.cluster.rabbitmq_docker_id))
-        instance.query("CREATE DATABASE test")
-        instance3.query("CREATE DATABASE test")
-
         yield cluster
 
     finally:
@@ -83,9 +80,12 @@ def rabbitmq_cluster():
 @pytest.fixture(autouse=True)
 def rabbitmq_setup_teardown():
     logging.debug("RabbitMQ is available - running test")
+    instance.query("CREATE DATABASE test")
+    instance3.query("CREATE DATABASE test")
     yield  # run test
     instance.query("DROP DATABASE test SYNC")
-    instance.query("CREATE DATABASE test")
+    instance3.query("DROP DATABASE test SYNC")
+    cluster.reset_rabbitmq()
 
 
 # Tests

--- a/tests/integration/test_storage_rabbitmq/test_failed_connection.py
+++ b/tests/integration/test_storage_rabbitmq/test_failed_connection.py
@@ -47,28 +47,12 @@ instance3 = cluster.add_instance(
 # Helpers
 
 
-def kill_rabbitmq(rabbitmq_cluster):
-    try:
-        p = subprocess.Popen(("docker", "stop", rabbitmq_cluster.rabbitmq_docker_id), stdout=subprocess.PIPE)
-        p.wait(timeout=30)
-        return p.returncode == 0
-    except Exception as ex:
-        print("Exception stopping rabbit MQ, will try forcefully", ex)
-        try:
-            p = subprocess.Popen(
-                ("docker", "stop", "-s", "9", rabbitmq_cluster.rabbitmq_docker_id), stdout=subprocess.PIPE
-            )
-            p.wait(timeout=30)
-            return p.returncode == 0
-        except Exception as e:
-            print("Exception stopping rabbit MQ forcefully", e)
-    finally:
-        time.sleep(4)
+def suspend_rabbitmq(rabbitmq_cluster):
+    rabbitmq_cluster.stop_rabbitmq_app()
 
 
-def revive_rabbitmq(rabbitmq_cluster):
-    p = subprocess.Popen(("docker", "start", rabbitmq_cluster.rabbitmq_docker_id), stdout=subprocess.PIPE)
-    p.wait(timeout=60)
+def resume_rabbitmq(rabbitmq_cluster):
+    rabbitmq_cluster.start_rabbitmq_app()
     rabbitmq_cluster.wait_rabbitmq_to_start()
 
 
@@ -80,9 +64,6 @@ def rabbitmq_cluster():
     try:
         cluster.start()
         logging.debug("rabbitmq_id is {}".format(instance.cluster.rabbitmq_docker_id))
-        instance.query("CREATE DATABASE test")
-        instance3.query("CREATE DATABASE test")
-
         yield cluster
 
     finally:
@@ -92,9 +73,12 @@ def rabbitmq_cluster():
 @pytest.fixture(autouse=True)
 def rabbitmq_setup_teardown():
     logging.debug("RabbitMQ is available - running test")
+    instance.query("CREATE DATABASE test")
+    instance3.query("CREATE DATABASE test")
     yield  # run test
     instance.query("DROP DATABASE test SYNC")
-    instance.query("CREATE DATABASE test")
+    instance3.query("DROP DATABASE test SYNC")
+    cluster.reset_rabbitmq()
 
 
 # Tests
@@ -167,8 +151,8 @@ def test_rabbitmq_restore_failed_connection_without_losses_1(rabbitmq_cluster):
     else:
         pytest.fail(f"Time limit of 180 seconds reached. The count is still 0.")
 
-    kill_rabbitmq(rabbitmq_cluster)
-    revive_rabbitmq(rabbitmq_cluster)
+    suspend_rabbitmq(rabbitmq_cluster)
+    resume_rabbitmq(rabbitmq_cluster)
 
     deadline = time.monotonic() + 180
     while time.monotonic() < deadline:
@@ -251,8 +235,8 @@ def test_rabbitmq_restore_failed_connection_without_losses_2(rabbitmq_cluster):
     else:
         pytest.fail(f"Time limit of 180 seconds reached. The count is still 0.")
 
-    kill_rabbitmq(rabbitmq_cluster)
-    revive_rabbitmq(rabbitmq_cluster)
+    suspend_rabbitmq(rabbitmq_cluster)
+    resume_rabbitmq(rabbitmq_cluster)
 
     # while int(instance.query('SELECT count() FROM test.view')) == 0:
     #    time.sleep(0.1)

